src/server/services/sgid.ts

```diff
@@ -0,0 +1,98 @@
+import SgidClient, { generatePkcePair } from '@opengovsg/sgid-client'
+
+interface SgidServiceOption {
+  clientId: string
+  clientSecret: string
+  privateKey: string
+  redirectUri: string
+  hostname: string
+}
+
+export const SGID_LOGIN_OAUTH_STATE = 'login'
+
+class SgidService {
+  // SGID client will be null for other domains like edu, health
+  // so we will need to handle cases for those as well, so that
+  // they continue to work in prod without sgid client being initialised
+  private sgidClient: SgidClient | null
+
+  constructor({
+    clientId,
+    clientSecret,
+    privateKey,
+    redirectUri,
+    hostname,
+  }: SgidServiceOption) {
+    try {
+      this.sgidClient = new SgidClient({
+        clientId,
+        clientSecret,
+        privateKey,
+        redirectUri,
+        hostname,
+      })
+    } catch (e) {
+      this.sgidClient = null
+    }
+  }
+
+  /**
+   * Fetches the url via sgid SDK.
+   */
+  authorizationUrl(): { url: string; codeVerifier: string; nonce?: string } {
+    if (!this.sgidClient) {
+      throw new Error('SGID client not initialised')
+    }
+    try {
+      const { codeChallenge, codeVerifier } = generatePkcePair()
+      const { url, nonce } = this.sgidClient.authorizationUrl({
+        state: SGID_LOGIN_OAUTH_STATE,
+        scope: ['openid', 'ogpofficerinfo.work_email'].join(' '),
+        codeChallenge,
+      })
+      return { url, codeVerifier, nonce }
+    } catch (e) {
+      throw new Error('Error retrieving url via sgid-client')
+    }
+  }
+
+  /**
+   * Fetches the token via sgid SDK.
+   */
+  async callback(code: string, nonce: string, codeVerifier: string) {
+    if (!code || !this.sgidClient)
+      throw Error(`code cannot be empty or sgid client not initialised`)
```

Comment:
```suggestion
      throw new Error(`code cannot be empty or sgid client not initialised`)
```

may I know why you decided to use it without the `new` keyword? functionally they [seem to be the same](https://www.geeksforgeeks.org/difference-between-throw-errormsg-and-throw-new-errormsg/) but I think it would be good to standardize

Reply:
Thanks for the catch!